libyul/optimiser/ValueConstraintBasedSimplifier.cpp

```diff
@@ -0,0 +1,384 @@
+*/
+/**
+ * Optimiser component that uses the simplification rules to simplify expressions.
+ */
+
+#include <libyul/optimiser/ValueConstraintBasedSimplifier.h>
+
+#include <libyul/optimiser/Semantics.h>
+
+#include <libyul/AsmData.h>
+#include <libyul/Utilities.h>
+#include <libyul/Exceptions.h>
+
+#include <libdevcore/CommonData.h>
+
+using namespace std;
+using namespace dev;
+using namespace yul;
+using namespace dev::solidity;
+
+namespace
+{
+class ConstraintDeduction: public boost::static_visitor<ValueConstraint>
+{
+public:
+	explicit ConstraintDeduction(map<YulString, ValueConstraint> const& _variableConstraints):
+		m_variableConstraints{_variableConstraints}
+	{}
+
+	ValueConstraint visit(Expression const& _expr)
+	{
+		return boost::apply_visitor(*this, _expr);
+	}
+
+	ValueConstraint operator()(Literal const& _literal)
+	{
+		return ValueConstraint::constant(valueOfLiteral(_literal));
+	}
+
+	ValueConstraint operator()(Identifier const& _identifier)
+	{
+		auto it = m_variableConstraints.find(_identifier.name);
+		if (it != m_variableConstraints.end())
+			return it->second;
+		else
+			return ValueConstraint{};
+	}
+
+	ValueConstraint operator()(FunctionCall const&)
+	{
+		return {};
+	}
+
+	ValueConstraint operator()(FunctionalInstruction const& _instr)
+	{
+		vector<Expression> const& args = _instr.arguments;
+		switch (_instr.instruction)
+		{
+		case dev::solidity::Instruction::ADD:
+			return visit(args.at(0)) + visit(args.at(1));
+		// TODO MUL
+		case dev::solidity::Instruction::SUB:
+			return visit(args.at(0)) - visit(args.at(1));
+		// TODO DIV, SDIV, MOD, SMOD, ADDMOD, MULMOD, EXP, SIGNEXTEND
+		case dev::solidity::Instruction::LT:
+			return visit(args.at(0)) < visit(args.at(1));
+		case dev::solidity::Instruction::GT:
+			return visit(args.at(1)) < visit(args.at(0));
+		// TODO SLT, SGT
+		case dev::solidity::Instruction::EQ:
+			return visit(args.at(1)) == visit(args.at(0));
+		case dev::solidity::Instruction::ISZERO:
+			return visit(args.at(0)) == ValueConstraint::constant(0);
+		case dev::solidity::Instruction::AND:
+			return visit(args.at(0)) & visit(args.at(1));
+		case dev::solidity::Instruction::OR:
+			return visit(args.at(0)) | visit(args.at(1));
+		// TODO XOR
+		case dev::solidity::Instruction::NOT:
+			return ~visit(args.at(0));
+		case dev::solidity::Instruction::BYTE:
+			// TODO Could be more specific.
+			return ValueConstraint::bitRange(7, 0);
+		case dev::solidity::Instruction::SHL:
+			return visit(args.at(1)) << visit(args.at(0));
+		case dev::solidity::Instruction::SHR:
+			return visit(args.at(1)) >> visit(args.at(0));
+		// TODO SAR
+		case dev::solidity::Instruction::ADDRESS:
+		case dev::solidity::Instruction::ORIGIN:
+		case dev::solidity::Instruction::CALLER:
+		case dev::solidity::Instruction::COINBASE:
+		case dev::solidity::Instruction::CREATE:
+		case dev::solidity::Instruction::CREATE2:
+			return ValueConstraint::bitRange(160, 0);
```

Comment:
```suggestion
			return ValueConstraint::bitRange(159, 0);
```

Reply:
Yep! already found that, too while adding tests.